pkg/policy/l4.go

```diff
@@ -292,27 +307,47 @@ func (l4 *L4Filter) ToMapState(policyOwner PolicyOwner, direction trafficdirecti
 	for cs, l7 := range l4.L7RulesPerSelector {
 		// Skip generating L3/L4 keys if L4-only key (for the same L4 port and
 		// protocol) has the same effect w.r.t. redirecting to the proxy or not,
-		// considering that L3/L4 key should redirect if L4-only key does. In
-		// summary, if have both L3/L4 and L4-only keys:
+		// considering that L3/L4 key should redirect if L4-only key does.
+		//
+		// Also consider that a deny policy should take precedence so that
+		// - l4-only deny overrides l3/l4 allows (and denies)
+		// - l3/l4 deny overrides l4-only allows (for that specific l3)
 		//
-		// L3/L4        L4-only      Skip generating L3/L4 key
-		// redirect     none         no
-		// no redirect  none         no
-		// redirect     no redirect  no   (this case tested below)
-		// no redirect  no redirect  yes  (same effect)
-		// redirect     redirect     yes  (same effect)
-		// no redirect  redirect     yes  (must redirect if L4-only redirects)
+		// In summary, if have both L3/L4 and L4-only keys:
 		//
-		// have wildcard?        this is a L3L4 key?  not the "no" case?
-		if l4.wildcard != nil && cs != l4.wildcard && !(l7 != nil && wildcardL7Policy == nil) {
+		//    Current L3/L4         L4-only (if any)    Skip generating L3/L4 key
+		//    redirect    allow     none        none    no
+		//    no redirect allow     none        none    no
+		//    no redirect deny      none        none    no
+		// ----------------------------------------------------------------------------------
+		// *) redirect    allow     no redirect allow   no   (this case tested below)
+		// ----------------------------------------------------------------------------------
+		//    redirect    allow     no redirect deny    yes  (deny takes precedence)
+		//    no redirect allow     no redirect allow   yes  (same effect)
+		//    no redirect allow     no redirect deny    yes  (deny takes precedence)
+		//    no redirect deny      no redirect allow   no   (deny takes precedence)
+		//    no redirect deny      no redirect deny    yes  (same effect)
+		//    redirect    allow     redirect    allow   yes  (same effect)
+		//    no redirect allow     redirect    allow   yes  (must redirect if L4-only redirects)
+		//    no redirect deny      redirect    allow   no   (deny takes precedence)
+		//
+		// have wildcard?        this is a L3L4 key?
+		isDeny := l7 != nil && l7.IsDeny
+		wildcardL7PolicyIsDeny := wildcardL7Policy != nil && wildcardL7Policy.IsDeny
+
+		if l4.wildcard != nil && cs != l4.wildcard &&
+			// not the "no" case marked with *)?
```

Comment:
When this comment is removed from the question it is answering, `not the "no" case` becomes unintelligible.

Surely we can just...?
```suggestion
			// Case marked with *)?
```

Reply:
I realized afterwards that maybe "not the no case" is referring to the right hand side column in the table above, so maybe it's trying to capture the complexity of all of those conditions? Still not exactly sure though.